File: src/main/java/com/qtech/bubbles/entity/GiantBubbleEntity.java
package com.qtech.bubbles.entity;

import com.qtech.bubbles.common.BubbleProperties;
import com.qtech.bubbles.common.entity.AbstractBubbleEntity;
import com.qtech.bubbles.common.entity.Attribute;
import com.qtech.bubbles.common.gametype.AbstractGameType;
import com.qtech.bubbles.common.random.BubbleRandomizer;
import com.qtech.bubbles.entity.types.EntityType;
import com.qtech.bubbles.environment.Environment;
import com.qtech.bubbles.init.Entities;

import java.awt.*;

/**
 * <h1>Bubble Entity.</h1>
 * One of the most important parts of the game.
 *
 * @see AbstractBubbleEntity
 */
@SuppressWarnings({"unused", "SameParameterValue"})
public class GiantBubbleEntity extends BubbleEntity {
    // Entity type.
    private static final EntityType<BubbleEntity> entityType = Entities.BUBBLE.get();

    public GiantBubbleEntity(AbstractGameType gameType) {
        super(gameType);
        this.addCollidable(Entities.PLAYER.get());
    }

    /**
     * <h1>Spawn Event Handler</h1>
     * On-spawn.
     *
     * @param pos         the position for spawn.
     * @param environment the environment to spawn in.
     */
    @Override
    public void onSpawn(Point pos, Environment environment) {
        BubbleRandomizer randomizer = this.gameType.getBubbleRandomizer();

        // Get random properties
        BubbleProperties properties = randomizer.getRandomProperties(environment.getGameType().getGameBounds(), gameType);

        // Bubble Type
        this.bubbleType = properties.getType();

        // Attributes.
        this.attributes.setBase(Attribute.ATTACK, bubbleType.getAttack(gameType, gameType.getBubbleRandomizer().getAttackRng()));
        this.attributes.setBase(Attribute.DEFENSE, bubbleType.getDefense(gameType, gameType.getBubbleRandomizer().getDefenseRng()));
        this.attributes.setBase(Attribute.SCORE_MULTIPLIER, bubbleType.getScore(gameType, gameType.getBubbleRandomizer().getScoreMultiplierRng()));

        // Dynamic values
        this.radius = properties.getRadius() * 4 + 80;
        this.speed = properties.getSpeed() / (Math.PI);
        this.attributes.setBase(Attribute.SPEED, (float) (properties.getSpeed() / (Math.PI)));
        this.baseRadius = properties.getRadius() * 4 + 80;
        this.attributes.setBase(Attribute.MAX_DAMAGE, properties.getDamageValue() * 4 + 80);
        this.damageValue = properties.getDamageValue() * 4 + 80;

        // Static values.
        this.bounceAmount = bubbleType.getBounceAmount();

        // Set velocity
        this.velX = -baseSpeed;

        bindEvents();
    }
}

File: src/main/kotlin/com/qtech/bubbles/entity/GiantBubbleEntity.kt
package com.qtech.bubbles.entity

import com.qtech.bubbles.common.entity.Attribute
import com.qtech.bubbles.common.gametype.AbstractGameMode
import com.qtech.bubbles.environment.Environment
import com.qtech.bubbles.init.Entities
import java.awt.Point

/**
 * <h1>Bubble Entity.</h1>
 * One of the most important parts of the game.
 *
 * @see AbstractBubbleEntity
 */
class GiantBubbleEntity(gameMode: AbstractGameMode) : BubbleEntity(gameMode) {
    /**
     * <h1>Spawn Event Handler</h1>
     * On-spawn.
     *
     * @param pos         the position for spawn.
     * @param environment the environment to spawn in.
     */
    override fun onSpawn(pos: Point?, environment: Environment?) {
        val randomizer = gameMode.bubbleRandomizer

        // Get random properties
        val properties = randomizer.getRandomProperties(environment!!.gameMode.gameBounds!!, gameMode)

        // Bubble Type
        bubbleType = properties.type

        // Attributes.
        attributeMap.set(Attribute.ATTACK, bubbleType.getAttack(gameMode, gameMode.bubbleRandomizer.attackRng))
        attributeMap.set(Attribute.DEFENSE, bubbleType.getDefense(gameMode, gameMode.bubbleRandomizer.defenseRng))
        attributeMap.set(Attribute.SCORE_MULTIPLIER, bubbleType.getScore(gameMode, gameMode.bubbleRandomizer.scoreMultiplierRng))

        // Dynamic values
        radius = properties.radius * 4 + 80
        speed = (properties.speed / Math.PI).toFloat()
        attributeMap.set(Attribute.SPEED, (properties.speed / Math.PI).toFloat())
        baseRadius = properties.radius * 4 + 80
        attributeMap.set(Attribute.MAX_DAMAGE, properties.damageValue * 4 + 80)
        damageValue = properties.damageValue * 4 + 80

        // Static values.
        bounceAmount = bubbleType.bounceAmount

        // Set velocity
        velocityX = (-baseSpeed)
        bindEvents()
    }

    init {
        this.addCollidable(Entities.PLAYER.get())
    }
}